# Day 3: No Matter How You Slice It
[link](https://adventofcode.com/2018/day/3)

## Part 1: Area of intersection

In [1]:
import numpy as np

def intersect_area(claims):
  area = np.zeros((1000, 1000), dtype=np.int16)
  for _, x, y, w, h in claims:
    area[x:x+w, y:y+h] += 1
  return np.sum(area > 1)

In [2]:
test_inp = [[1,1,3,4,4], [2,3,1,4,4], [3,5,5,2,2]]
test_expected = 4
test_actual = intersect_area(test_inp)
assert test_actual == 4, f'Expected {test_expected}, got {test_actual}'

In [3]:
import re

with open('03 input.txt', 'r') as file:
  inp = [list(map(int, re.findall(r'[\+-]?\d+', line))) for line in file]

In [4]:
part_1_answer = intersect_area(inp)
assert part_1_answer == 111266

Right answer: `111266`

## Part 2: ID of a claim that does not overlap with others

In [5]:
def non_overlapping_ids(claims):
  area = np.zeros((1000, 1000), dtype=np.int16)
  for id,x,y,w,h in claims:
    claim_area = area[x:x+w, y:y+h]
    if np.all(claim_area == 0):
      claim_area[:] = id
    else:
      overlapped_ids = np.unique(claim_area[claim_area > 0])
      for overlapped_id in overlapped_ids:
        area[area == overlapped_id] = -1
      claim_area[:] = -1
#   print(area[:10,:10])
  return np.unique(area[area > 0])

In [6]:
# test_inp_2 = [[1,1,3,4,4], [2,3,1,4,4], [3,5,5,2,2], [4,6,7,2,2]]
test_inp_2 = [[7,4,4,2,2], [2,5,6,4,4]]

test_result_2 = non_overlapping_ids(test_inp_2)

print(', '.join(map(str, test_result_2)))
# assert sorted(test_result_2.tolist()) == [3, 4]

2, 7


In [7]:
non_overlapping_ids(inp)

array([266], dtype=int16)

**Wrong answers**
- `117` - too low. Reason: used NumPy int8 data type (max value 127) for IDs, which reached 1000 and higher.

**Correct answer:** `266`